app/src/main/java/org/listenbrainz/android/util/JobQueue.kt

```diff
@@ -0,0 +1,169 @@
+                            // Remove token from map if empty.
+                            if (jobList.isEmpty()) {
+                                jobsMap.remove(queueJob.token)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    fun post(
+        token: Any? = null,
+        context: CoroutineContext = dispatcher,
+        block: suspend CoroutineScope.() -> Unit
+    ) {
+        scope.launch {
+            val queueJob = createQueueJob(token, context, block)
+            addJobToMap(queueJob)
+            sendJobToQueue(queueJob)
+        }
+    }
+    
+    fun postDelayed(
+        delayMillis: Long,
+        token: Any? = null,
+        context: CoroutineContext = dispatcher,
+        block: suspend CoroutineScope.() -> Unit
+    ) {
+        scope.launch {
+            val queueJob = createQueueJob(token, context, block)
+            addJobToMap(queueJob)
+            delay(delayMillis)
+            sendJobToQueue(queueJob)
+        }
+    }
+    
+    /** Removes all the delayed posts with [token] identifier.
+     * @param token if null, all jobs will be cancelled.*/
+    fun removePosts(token: Any?) {
+        scope.launch(dispatcher) {
+            if (token == null) {
+                /** Mark all jobs as cancelled. */
+                mapLock.withLock {
+                    jobsMap.forEach {
+                        val jobList = it.value
+                        jobList.forEach { queueJob ->
+                            queueJob.cancelled.set(true)
+                        }
+                    }
+                }
+            } else {
+                /** Mark job with unique token as cancelled. */
+                mapLock.withLock {
+                    val jobList = jobsMap[token]
+                    jobList?.forEach { queueJob ->
+                        queueJob.cancelled.set(true)
+                    }
+                }
+            }
+        }
+    }
+    
+    private fun createQueueJob(
+        token: Any?,
+        context: CoroutineContext,
+        block: suspend CoroutineScope.() -> Unit
+    ): QueueJob {
+        val job = scope.launch(context, CoroutineStart.LAZY) {
+            if (isActive) {
+                block()
+            }
+        }
+        
+        return QueueJob(token, job, AtomicBoolean(false))
+    }
+    
+    private suspend fun addJobToMap(queueJob: QueueJob) {
+        mapLock.withLock {
+            if (jobsMap.containsKey(queueJob.token)) {
+                // We should not need to do null checks since we are using locks but still
+                // are for safety.
+                jobsMap[queueJob.token]?.add(queueJob)
+            } else {
+                jobsMap[queueJob.token] = LinkedList<QueueJob>().apply { add(queueJob) }
+            }
+        }
+    }
+    
+    private suspend fun sendJobToQueue(queueJob: QueueJob) {
+        syncLock.withLock {
```

Comment:
Remove as this is redundant. queue.send is already synchronous

Reply:
Noice catch

Reply:
@07jasjeet I see that you didn't remove suspend from this function even after the update. Also, since this a one-liner anyway, we can omit keeping this function and use the logic directly at places